lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -171,439 +172,348 @@ class ConvertAtenFlattenUsingIntsOp
-  // all the dimensions in `outputShape`.
-  static void calculateSingleDynamicSize(MutableArrayRef<int64_t> inputShape,
-                                         MutableArrayRef<int64_t> outputShape) {
-    if (inputShape.empty() || outputShape.empty())
-      return;
-    int64_t inputDynamicDimCount = llvm::count(inputShape, kUnknownSize);
-    int64_t outputDynamicDimCount = llvm::count(outputShape, kUnknownSize);
-    if (inputDynamicDimCount + outputDynamicDimCount != 1)
+  // This function assumes that all the dimensions in `inShape` map to all the
+  // dimensions in `outShape`.
+  static void tryFillBothDirections(MutableArrayRef<int64_t> inShape,
+                                    MutableArrayRef<int64_t> outShape) {
+
+    int64_t inDynamicDimCount = llvm::count(inShape, kUnknownSize);
+    int64_t outDynamicDimCount = llvm::count(outShape, kUnknownSize);
+
+    if (inDynamicDimCount == 0 && outDynamicDimCount == 0)
       return;
 
-    int64_t inputProduct = productReduce(inputShape);
-    int64_t outputProduct = productReduce(outputShape);
+    // Try and fill in the gaps in `outShape` using `inShape`.
+    else if (inDynamicDimCount == 0)
+      tryFillOneDirection(inShape, outShape);
 
-    if (inputDynamicDimCount == 1) {
-      inputProduct /= kUnknownSize;
-      *llvm::find(inputShape, kUnknownSize) = outputProduct / inputProduct;
-    } else {
-      outputProduct /= kUnknownSize;
-      *llvm::find(outputShape, kUnknownSize) = inputProduct / outputProduct;
-    }
+    // Try and fill in the gaps in `inShape` using `outShape`.
+    else if (outDynamicDimCount == 0)
+      tryFillOneDirection(outShape, inShape);
   }
 
-  // Gets the shapes of the input and output tensors, making a best-effort
+  // Get the shapes of the input and output tensors, making a best-effort
   // attempt to extract static shape information given the inputs to
-  // `aten.view`.
+  // `aten.view`. That is, try to have as few kUnknownSize entries as possible.
   static std::pair<SmallVector<int64_t>, SmallVector<int64_t>>
-  getInputAndOutputShape(Value inputTorchTensor,
-                         SmallVector<Value> outputSizeTorchInt) {
-    SmallVector<int64_t> inputShape(
-        inputTorchTensor.getType().cast<BaseTensorType>().getSizes());
-    SmallVector<int64_t> outputShape(outputSizeTorchInt.size(), kUnknownSize);
-    for (auto [outputDim, outputDimSize] :
-         llvm::enumerate(outputSizeTorchInt)) {
-      int64_t inputDim;
-      int64_t outputDimSizeInt;
-      // Match torch.aten.size.int(inputTensor, inputDim) with constant inputDim
-      if (matchPattern(outputDimSize,
-                       m_TorchTensorSizeInt(inputTorchTensor, &inputDim))) {
-        outputShape[outputDim] = inputShape[inputDim];
-      } else if (matchPattern(outputDimSize,
-                              m_TorchConstantInt(&outputDimSizeInt))) {
-        if (outputDimSizeInt != -1) {
```

Comment:
Why are we able to drop this check? If a dimension size is `-1` then it needs to be inferred from the rest of the shape of the view, meaning it either can be statically resolved as an earlier step or we need to keep it dynamic.